File: Core/src/main/kotlin/jokatu/game/input/InputDeserialiser.java
package jokatu.game.input;

import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.Map;

/**
 * This takes JSON from the client and turns it into input.
 * @param <I> the type of the {@link Input} to output
 */
public abstract class InputDeserialiser<I extends Input> {

	@NotNull
	public abstract I deserialise(@NotNull Map<String, Object> json) throws DeserialisationException;

	@NotNull
	protected Object getMandatoryKeyValue(@NotNull String keyName, @NotNull Map<String, Object> json) throws DeserialisationException {
		if (!json.containsKey(keyName)) {
			throw new DeserialisationException(json, MessageFormat.format("Did not contain the key ''{0}''.", keyName));
		}
		Object value = json.get(keyName);
		if (value == null) {
			throw new DeserialisationException(json, MessageFormat.format("The value for the key ''{0}'' was null.", keyName));
		}
		return value;
	}

	protected <T> T getMandatoryKeyValueOfType(
			@NotNull Class<T> type,
			@NotNull String keyName,
			@NotNull Map<String, Object> json
	) throws DeserialisationException {
		return castValue(type, keyName, getMandatoryKeyValue(keyName, json), json);
	}

	@NotNull
	protected <T> T castValue(
			@NotNull Class<T> type,
			@NotNull String keyName,
			@NotNull Object value,
			@NotNull Map<String, Object> json
	) throws DeserialisationException {
		if (!type.isInstance(value)) {
			throw new DeserialisationException(
					json,
					MessageFormat.format(
							"The value for ''{0}'' was not a {1}: instead, was the {2} ''{3}''",
							keyName,
							type.getSimpleName(),
							value.getClass().getSimpleName(),
							value
					)
			);
		}
		return type.cast(value);
	}
}

File: Core/src/main/kotlin/jokatu/game/input/InputDeserialiser.kt
package jokatu.game.input

/**
 * This takes JSON from the client and turns it into input.
 * @param <I> the type of the [Input] to output
 */
abstract class InputDeserialiser<I : Input> {

	@Throws(DeserialisationException::class)
	abstract fun deserialise(json: Map<String, Any>): I

	@Throws(DeserialisationException::class)
	protected fun getMandatoryKeyValue(keyName: String, json: Map<String, Any>): Any {
		if (!json.containsKey(keyName)) {
			throw DeserialisationException(json, "Did not contain the key '$keyName'.")
		}
		return json[keyName]
				?: throw DeserialisationException(json, "The value for the key '$keyName' was null.")
	}

	@Throws(DeserialisationException::class)
	protected fun <T> getMandatoryKeyValueOfType(
			type: Class<T>,
			keyName: String,
			json: Map<String, Any>
	): T {
		return castValue(type, keyName, getMandatoryKeyValue(keyName, json), json)
	}

	@Throws(DeserialisationException::class)
	protected fun <T> castValue(
			type: Class<T>,
			keyName: String,
			value: Any,
			json: Map<String, Any>
	): T {
		if (!type.isInstance(value)) {
			throw DeserialisationException(
					json,
					"The value for '$keyName' was not a ${type.simpleName}: instead, was the ${value.javaClass.simpleName} '$value'."
			)
		}
		return type.cast(value)
	}
}
